File: graph/graph_via_adjacent_list.c
/**
 * @file graph_via_adjacent_list.c
 * @brief A graph structure implementation via adjacent list apporach instead of adjacent matrix.
 */
//

#include <complex.h>
#include <stdio.h>
#include <stdlib.h>
#include <memory.h>
#include <stdbool.h>

#define MAX_VERTICES 100

// Edge structure (Represent a single edge in the graph)
typedef struct edge {
    unsigned int sourceEdge;                    // The source vertex of the edge
    unsigned int destinationEdge;               // The destination vertex of the edge
    bool isDirected;                            // Directed or undirected edge (If directed, source -> destination, otherwise source <-> destination)
} Edge;

// A single node in the graph (Represent a single vertex in the graph)
typedef struct graphNode {
    unsigned int vertex;                        // The number of the vertex
    struct graphNode* next;                     // The next node in the linked list
} GraphNode;

// Graph structure
typedef struct graphViaAdjacentList {
    unsigned int vertices;                      // The number of vertices in the graph
    unsigned int edges;                         // The number of edges in the graph
    GraphNode *adjacentList[MAX_VERTICES];      // The adjacent list of the graph (Each vertex has a linked list of the connected vertices)
} GraphViaAdjacentList;

// Create a graph via adjacent list
GraphViaAdjacentList* createGraphViaAdjacentList() {
    GraphViaAdjacentList* graph = (GraphViaAdjacentList*) malloc(sizeof(GraphViaAdjacentList));
    graph->vertices = 0;
    graph->edges = 0;
    memset(graph->adjacentList, 0, sizeof(graph->adjacentList));
    return graph;
}

// Add an edge to the graph
void addEdgeToGraphViaAdjacentList(GraphViaAdjacentList* graph, Edge* edge) {
    if (edge->sourceEdge < MAX_VERTICES && edge->destinationEdge < MAX_VERTICES) {
        GraphNode* node = (GraphNode*) malloc(sizeof(GraphNode));
        node->vertex = edge->destinationEdge;
        node->next = graph->adjacentList[edge->sourceEdge];
        graph->adjacentList[edge->sourceEdge] = node;
        graph->edges++;

        if (!edge->isDirected) {
            // If the edge is undirected, add the reverse edge
            GraphNode* reverseNode = (GraphNode*) malloc(sizeof(GraphNode));
            reverseNode->vertex = edge->sourceEdge;
            reverseNode->next = graph->adjacentList[edge->destinationEdge];
            graph->adjacentList[edge->destinationEdge] = reverseNode;
            graph->edges++;                     // Consider the reverse edge as well
        }
    } else {
        // Invalid edge source or destination
        fprintf(stderr, "Invalid edge source or destination\n");
    }
}

// Remove an edge from the graph
// Find the edge in the source vertex and remove it
// Removing procedure is almost similar to handle the removing elements in the linked list
void removeEdgeFromGraphViaAdjacentList(GraphViaAdjacentList* graph, Edge* edge) {
    if (edge->sourceEdge < MAX_VERTICES && edge->destinationEdge < MAX_VERTICES) {
        GraphNode* node = graph->adjacentList[edge->sourceEdge];                // Find the edge in the source vertex (This is the linked list of the edge)
        GraphNode* previousNode = NULL;                                         // Keep track of the previous node
        while (node != NULL) {
            // Find the edge in the source vertex
            if (node->vertex == edge->destinationEdge) {
                // Remove the edge (we found the edge in the source vertex)
                if (previousNode == NULL) {
                    // The edge is the first edge in the list
                    graph->adjacentList[edge->sourceEdge] = node->next;
                } else {
                    // The edge is not the first edge in the list
                    previousNode->next = node->next;
                }
                free(node);
                graph->edges--;
                break;
            }
            // We didn't find the edge yet, continue to the next node
            previousNode = node;
            node = node->next;
        }

        // If the edge is undirected, remove the reverse edge as well
        if (!edge->isDirected) {
            // If the edge is undirected, remove the reverse edge
            node = graph->adjacentList[edge->destinationEdge];
            previousNode = NULL;
            while (node != NULL) {
                if (node->vertex == edge->sourceEdge) {
                    // Remove the edge (we found the edge in the destination vertex)
                    if (previousNode == NULL) {
                        graph->adjacentList[edge->destinationEdge] = node->next;
                    } else {
                        previousNode->next = node->next;
                    }
                    free(node);
                    graph->edges--;
                    break;
                }
                previousNode = node;
                node = node->next;
            }
        }
    } else {
        // Invalid edge source or destination
        fprintf(stderr, "Invalid edge source or destination\n");
    }
}

// Print the graph
void printGraphViaAdjacentList(GraphViaAdjacentList* graph) {
    for (unsigned int vertex = 0; vertex < MAX_VERTICES; vertex++) {
        GraphNode* node = graph->adjacentList[vertex];
        if (node != NULL) {
            printf("Vertex %d is connected to: ", vertex);
            while (node != NULL) {
                printf("%d ", node->vertex);
                node = node->next;
            }
            printf("\n");
        }
    }
}

// Free the graph memory (Free the graph and all the nodes in the graph)
void freeGraphViaAdjacentList(GraphViaAdjacentList *graph) {
    for (unsigned int vertex = 0; vertex < MAX_VERTICES; vertex++) {
        GraphNode* node = graph->adjacentList[vertex];
        while (node != NULL) {
            // Traverse the linked list and free the memory
            GraphNode* temp = node;
            node = node->next;
            free(temp);
        }
    }
    printf("Freed the graph memory\n");
}

// Main function
int main(void) {
    GraphViaAdjacentList* graph = createGraphViaAdjacentList();

    addEdgeToGraphViaAdjacentList(graph, &(Edge){0, 1, false});
    addEdgeToGraphViaAdjacentList(graph, &(Edge){0, 2, false});
    addEdgeToGraphViaAdjacentList(graph, &(Edge){1, 2, false});
    addEdgeToGraphViaAdjacentList(graph, &(Edge){2, 3, false});
    addEdgeToGraphViaAdjacentList(graph, &(Edge){3, 4, false});
    addEdgeToGraphViaAdjacentList(graph, &(Edge){4, 0, false});
    
    printGraphViaAdjacentList(graph);
    printf("\n");

    removeEdgeFromGraphViaAdjacentList(graph, &(Edge){1, 2, false});
    printGraphViaAdjacentList(graph);

    return 0;
}

// Vertex 0 is connected to: 4 2 1 
// Vertex 1 is connected to: 2 0
// Vertex 2 is connected to: 3 1 0
// Vertex 3 is connected to: 4 2
// Vertex 4 is connected to: 0 3
// 
// Vertex 0 is connected to: 4 2 1
// Vertex 1 is connected to: 0
// Vertex 2 is connected to: 3 0
// Vertex 3 is connected to: 4 2
// Vertex 4 is connected to: 0 3

File: graph/graph_via_adjacent_list.py
# file: graph_via_adjacent_list.py
# brief: A general implementation of a graph(data structure) in Python with class
#        Basically, this is implemented using an adjacency list.

from typing import Dict, List, Tuple

class Graph:
    def __init__(self) -> None:
        self.graph: Dict[str, List[str]] = {}

    # Add a new vertex to the graph
    def add_vertex(self, vertex: str) -> None:
        if not vertex:
            raise ValueError("Vertex cannot be empty")
        if vertex not in self.graph:
            self.graph[vertex] = []
        else:
            raise ValueError("Vertex already exists")
        
    # Add a new edge to the graph
    # We have an edge from source_vertex to destination_vertex (directed graph)
    def add_edge(self, source_vertex: str, destination_vertex: str, directed: bool) -> None:
        if not source_vertex or not destination_vertex:
            raise ValueError("Source and destination vertices cannot be empty")
        self.graph[source_vertex].append(destination_vertex)
        if not directed:
            # If the graph is undirected, add an edge from destination_vertex to source_vertex
            self.graph[destination_vertex].append(source_vertex)

    # Remove a vertex from the graph
    def remove_vertex(self, vertex: str) -> None:
        if not vertex:
            raise ValueError("Vertex cannot be empty")
        if vertex in self.graph:
            del self.graph[vertex]

            # Remove the target vertex from all other vertices' adjacency lists
            for vertices in self.graph:
                if vertex in self.graph[vertices]:
                    self.graph[vertices].remove(vertex)
        else:
            raise ValueError("Vertex does not exist")
        
    # Remove an edge from the graph
    def remove_edge(self, source_vertex: str, destination_vertex: str) -> None:
        if not source_vertex or not destination_vertex:
            raise ValueError("Source and destination vertices cannot be empty")
        if source_vertex in self.graph:

            # Remove the edge from source_vertex to destination_vertex
            if destination_vertex in self.graph[source_vertex]:
                self.graph[source_vertex].remove(destination_vertex)
            else:
                raise ValueError("Edge does not exist")

    # Print the graph
    def print_graph(self) -> None:
        for vertex in self.graph:
            print(f"{vertex} -> {self.graph[vertex]}")

    # Get the number of vertices in the graph
    def get_num_vertices(self) -> int:
        return len(self.graph)
    
    # Get the number of edges in the graph
    def get_num_edges(self) -> int:
        num_edges = 0
        for vertex in self.graph:
            num_edges += len(self.graph[vertex])
        return num_edges
    
# Now, let's use the graph.
graph = Graph()

graph.add_vertex("A")
graph.add_vertex("B")
graph.add_vertex("C")
graph.add_vertex("D")
graph.add_vertex("E")
graph.add_vertex("F")

graph.add_edge(source_vertex="A", destination_vertex="B", directed=True)
graph.add_edge(source_vertex="A", destination_vertex="C", directed=True)
graph.add_edge(source_vertex="B", destination_vertex="D", directed=False)
graph.add_edge(source_vertex="C", destination_vertex="E", directed=True)
graph.add_edge(source_vertex="D", destination_vertex="E", directed=True)
graph.add_edge(source_vertex="E", destination_vertex="F", directed=False)
graph.add_edge(source_vertex="F", destination_vertex="A", directed=True)

graph.print_graph()

print(f"Number of vertices: {graph.get_num_vertices()}")
print(f"Number of edges: {graph.get_num_edges()}")
print("\n")

graph.remove_vertex("A")
graph.remove_edge(source_vertex="B", destination_vertex="D")
graph.print_graph()

print(f"Number of vertices: {graph.get_num_vertices()}")
print(f"Number of edges: {graph.get_num_edges()}")

# A -> ['B', 'C']
# B -> ['D']
# C -> ['E']
# D -> ['B', 'E']
# E -> ['F']
# F -> ['E', 'A']
# Number of vertices: 6
# Number of edges: 9
# 
# 
# B -> []
# C -> ['E']
# D -> ['B', 'E']
# E -> ['F']
# F -> ['E']
# Number of vertices: 5
# Number of edges: 5